Add migrations for factory files, config, and various other files

derrickmehaffy opened the pull request:
## What is this

Basically we were not creating the factory files for controllers and services and for many users who had lots of content-types in v3 it was extremely manual and monotonous process to literally go through and copy paste then change the UID.

I have added the factory file for routes but I see code-mods is dumping the legacy conversion and I think we need to discuss further before I touch the routes migration as as I understand this is  also used for plugins.

My opinion is that we need two different route conversions; one for plugins and one for applications.

## To-Do List

- [x] Build [Liquid.js](https://liquidjs.com/) templates for all the new files
- [x] Create factory routes file
- [x] Create factory controllers file
- [x] Create factory service file
- [x] Create default v4 config files (and copy the existing config to a safe location)
- [x] Pick the proper database config file template based on v3 database
- [x] Generate `./src/admin/*` example files that currently break 4.3.x
- [x] Create new .env (and copy the existing to a safe location)
- [x] Fix tests

## Misc unrelated fixes

- [x] Fix admin relation conversion

Signed-off-by: Derrick Mehaffy <derrickmehaffy@gmail.com>

derrickmehaffy commented:
Example of the current output based on this PR: 
![image](https://user-images.githubusercontent.com/8593673/185650676-8ff31d47-3c72-4c3c-b59f-c11429296ffc.png)

While this migration "works" those routes are now static and I think that for normal APIs we should be using this template: 

https://github.com/strapi/codemods/blob/feature/migrateCoreContServ/lib/v4/migration-helpers/update-api-folder-structure/templates/core-router.liquid

markkaylor reviewed (commented):
Hey @derrickmehaffy I really appreciate this PR and I think it is a good start, but one problem I notice (and I think it might really complicate things 😬 ) is that this approach will not consider custom code.

For example in a controller, if a user created actions or extended the core actions none of that will be migrated it will just be replaced with what the factory provides. 

v3 Controller
![Screenshot 2022-08-22 at 09 45 33](https://user-images.githubusercontent.com/26598053/185867959-40515210-e40a-4633-99f9-c16627818bdf.png)

(after migration) 

Expected (more or less):
![Screenshot 2022-08-22 at 09 59 26](https://user-images.githubusercontent.com/26598053/185869998-5a9939fc-87b7-487d-9f58-590d0d2b2fbd.png)


Received:
![Screenshot 2022-08-22 at 09 58 11](https://user-images.githubusercontent.com/26598053/185869758-8a31350f-ff9f-4de0-bde9-e432c5aa68a3.png)

We would need to detect all this custom code and then pass it to the callback as [described in the docs](https://docs.strapi.io/developer-docs/latest/development/backend-customization/controllers.html#adding-a-new-controller). I'm not 100% confident it is even possible/reliable but if it is I think `jscodeshift` would probably be the tool to use to detect custom code and modify the file.

Another option would be to detect custom code, if none is found use the factories, otherwise use the current setup.

edit: Or I guess we could just pass everything to the callback and leave cleanup to the user. 

Otherwise I think the PR looks good and I agree that we should consider separating scripts for applications and plugins

derrickmehaffy commented:
The problem becomes @markkaylor that the user is gonna have to rewrite these custom controllers and services anyway and this saves a ton of time from forcing the user to manually copy/paste the factories for router, controller, and service for all of the content-types they -didn't modify-.

We could copy the old ones into another file (chatted with @alexandrebodin, @fdel-car, and @PaulBratslavsky in Alex's AMA today about this) but even then, that is limited.

The example use-case that I designed this PR around is a customer with roughly 180 content-types and only about 14 of them have custom controllers/services and the other 160-something don't, they had to manually copy paste 480 files and change a single UID in all of them by hand.

markkaylor commented:
Ok sounds good to me, It definitely simplifies things if we are ok with the fact that users will have to migrate custom code by hand.

> I see code-mods is dumping the legacy conversion

Not sure I understand exactly what you mean here, but all v3 files are stored in a v3 folder after running the script so the code for v3 controllers, services, and routes is still accessible after migration.

markkaylor commented:
For routes I think we could just leave it as is for plugins and create a new file `update-application-routes` that will implement the factories. Then use it in `/lib/v4/migration-helpers/update-api-folder-structure/index.js`. WDYT? Feel free to ping me on slack if you want to chat more about this on discord.

derrickmehaffy commented:
@markkaylor I did as you mentioned, basically there is a new router migration since we copy the old ones.

I also went the extra step here and am doing the same thing for config (partially done) but I need to check into figuring out which database the v3 project is using first so I can use the proper template.

I'd also like to do something with the .env so that some random keys can be properly generated in a `.env` file but I'll work on that hopefully tmrw or next week.

markkaylor commented on lib/v4/migration-helpers/update-api-folder-structure/index.js:
```diff
@@ -6,27 +6,39 @@ const { resolve, join, basename } = require('path');
   try {
-    // Copy the folder to a v3 folder for safe keeping
+    // Copy the config folder to the v3 folder for safe keeping
+    await fs.move(configPath, join(strapiAppPath, 'v3', 'config'));
```
I'm not sure this should be done here since it isn't related to the api.

I don't think it should be moved because it will still be at the root of the application, just copied to v3 for safe keeping

Could we do this in `update-application-folder-structure`, you probably could create the folder there if it doesn't exist instead of in `update-application-config.js`

Something like:

```js
    const configPath = join(strapiAppPath, 'config')
    const exists = await fs.exists(configPath);
    if (!exists) {
      await fs.copy(configPath, join('v3', 'config'));
    } else {
      // create config path
      await fs.mkdir(configPath);
    }
```

derrickmehaffy commented:
Updated to follow your feedback @markkaylor and also added a very dirty database check (using `.search`) to try and determine the database client to use on generating a fresh database config. I'm not sure if there is a better/cleaner way.

One thing that has been bugging me and I'm not sure if it's me or something else in the script but the codemods folder as I run it generates/copies the this here and not in the project repo: 

![image](https://user-images.githubusercontent.com/8593673/187549414-af4bf60b-2280-40c9-ab34-43581ec5c4f9.png)

I'm like 99% sure that shouldn't exist but let me know.

---

If this becomes validated thus far then I will finish fixing up the tests to match the new logic.

markkaylor reviewed (changes requested):
Nice work Derrick! Looking good. I just left a couple more small comments, and yeah we will need to fix the tests.

For the `extensions`, on my end:
- The extensions are moved for v4 to `src` which I believe is what is expected [according to the docs](https://docs.strapi.io/developer-docs/latest/development/plugins-extension.html)
- The extensions are also copied to the root of the v3 directory which I believe is what is expected as well...

Let me know if I'm missing something or if you don't have the same on your end

markkaylor commented:
Oh I just noticed in your screenshot the v3/extensions folder is inside the codemod repo 🤔 that's strange, I don't have that issue

derrickmehaffy commented on lib/v4/utils/strapi-packages.js:
```diff
@@ -1,42 +1,43 @@
 const strapiPackages = {
   strapi: '@strapi/strapi',
-  'strapi-admin': '@strapi/admin',
+  'strapi-admin': toBeRemoved,
```
It doesn't need to be set in the package.json, the `@strapi/strapi` imports it.

Trying to clean up the users package.json to match what a fresh project mimics

markkaylor replied:
yeah you are right, my bad I actually thought this was `@strapi/strapi` 🤦‍♂️

derrickmehaffy commented:
> Oh I just noticed in your screenshot the v3/extensions folder is inside the codemod repo 🤔 that's strange, I don't have that issue

How do you run code mods locally? I just did a `node bin/cli.js` so I think it has to do with an early on move or copy that isn't using the app path

markkaylor commented:
> How do you run code mods locally? I just did a node bin/cli.js so I think it has to do with an early on move or copy that isn't using the app path

I do the same using a relative path from my test project `../../somewhere/codemod/bin/cli.js migrate:application .`

derrickmehaffy commented:
> > How do you run code mods locally? I just did a node bin/cli.js so I think it has to do with an early on move or copy that isn't using the app path
> 
> I do the same using a relative path from my test project `../../somewhere/codemod/bin/cli.js migrate:application .`

So maybe it depends on where you call the CLI as I usually do it from within the codemods repo and pass the absolute path to my strapi project like `node bin/cli.js migrate:application /path/to/project`

derrickmehaffy commented:
Per mark via internal slack "looks good just needs tests fixed".

I'll try to get that finished up before tmrw. We do have a few users/customers testing this branch right now and I have a call with one of them later today.

PaulBratslavsky commented:
Thank you all for working on this.

PaulBratslavsky reviewed (approved):
Tested and works.

derrickmehaffy commented:
@marcoautiero can you try again as this should handle the .env issue regardless if the user had one existing 

Basically if there was a v3 `.env` the codemods loads it, parses out the database vars, and uses them to construct the new `.env` while copying the old one to the v3 folder.

It also uses the exact same logic we use in the monorepo to generate the various secrets.

remidej commented on lib/v4/migration-helpers/update-api-folder-structure/templates/core-service.liquid:
```diff
@@ -0,0 +1,9 @@
+'use strict';
+
+/**
+ * {{ id }} service
```
I think these commands would look nicer if we used the display name instead of id

derrickmehaffy replied:
Oh I was basing it on what we currently do with the existing generators I think :thinking: Though I do agree with you, it might look weird when there are differences between migrated files vs new ones.

remidej reviewed (commented):
I haven't reviewed the code, but found no issues migrating a simple v3 app with a couple of content types. The routes, controllers and services get created as expected 👌
